[Autograd] Use in-place input accumulation fast path for dense Tensors.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88339

There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

Commit:
[Autograd] Use in-place input accumulation fast path for dense Tensors.

There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

[ghstack-poisoned]

ngimel commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -86,13 +86,12 @@ static void accumulate(
       buffer[pos] = var + old_var;
     }
+  } else if (
+      old_var.is_contiguous() && old_var.use_count() == 1 &&
```
I don't think we want to take this path when both `var` and `old_var` are sparse, who even knows if we have working inplace sparse-sparse addition, and in any case it's a conceptually weird operation, so you might want to push this case to the `else` branch.

robieta replied:
If `old_var` is sparse it will get hooked by the first part of the conditional. (Which as a `!var.is_sparse()` check on its steal check.) Maybe I should add a cautionary comment?

ngimel replied:
Ah yeah that would be good, thanks for checking!

robieta commented:
Looks like I also need a `!old_var._is_zerotensor()` check. @anjali411 is there a more general mutability check that I should use?

Commit:
Update on "[Autograd] Use in-place input accumulation fast path for dense Tensors."


There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/input_buffer.cpp:
```diff
@@ -86,13 +86,12 @@ static void accumulate(
+  } else if (
+      old_var.is_contiguous() && !old_var._is_zerotensor() &&
+      old_var.use_count() == 1 && old_var.storage().use_count() == 1) {
+    buffer[pos] = old_var.add_(var);
```
I think there are a few things wrong here that need updating. In particular looking at the logic for AccumulateGrad show a lot of possible failure cases here: https://github.com/pytorch/pytorch/blob/0fa23663ccd5350469c95615ddb7d2fd2a88abe3/torch/csrc/autograd/functions/accumulate_grad.h#L56

- We need to check that grad mode is disabled. This inplace might be invalid if old_var is saved by autograd
- You need to guard for `.is_sparse_csr()` which is not covered by `.is_sparse()`
- We want is_non_overlapping_and_dense on top of is_contiguous as we really don't want weird memory overlapping Tensors to be written inplace.
- Why the special case on zero_tensor here?

ngimel replied:
>is_non_overlapping_and_dense on top of is_contiguous

instead of is contiguous? Yeah that's reasonable. IIRC it errors out on sparse tensors though, so the logic will need to become more tortured

robieta replied:
FWIW I'm also updating the path where `var` is used, so there will be sparse checks before any other checks.

albanD replied:
> Should that be GradMode::is_enabled() or a property of old_var?

GradMode::is_enabled() because it might not be ok to modify old_var even if it doesn't require gradients.
Note that it could still be invalid when grad mode is disabled but that should be unlikely enough here that we're ok.

Commit:
Update on "[Autograd] Use in-place input accumulation fast path for dense Tensors."


There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

robieta commented:
Updated with more rigorous checks and comments.

robieta commented:
The latest failure is `torch.autograd.gradcheck.GradcheckError: While computing batched gradients, got: Cannot access storage of BatchedTensorImpl`. Nested Tensor, sparse Tensor, ZeroTensor, BatchedTensor etc. will all blow up if various methods are called and it's virtually impossible to get everything right, much less future proof it. I'll fix the unit test failures, but I'm increasingly coming to the conclusion that we need to add something like `is_plain_old_tensor()` to Tensor.

zou3519 commented:
there's an isTensorSubclassLike check for this: https://github.com/pytorch/pytorch/blob/ef4ce6d4c6ce1bd5ec26d7f6f71f1c053da46945/aten/src/ATen/TensorSubclassLikeUtils.h#L51

We probably want to exclude tensors that check isTensorSubclassLike True from this path

robieta commented:
> there's an isTensorSubclassLike check for this, though it is not completely comprehensive yet:
> 
> https://github.com/pytorch/pytorch/blob/ef4ce6d4c6ce1bd5ec26d7f6f71f1c053da46945/aten/src/ATen/TensorSubclassLikeUtils.h#L51
> 
> We probably want to exclude tensors that check isTensorSubclassLike True from this fast path

This is AWESOME! (CC @chaekit, @aaronenyeshi, @slgong-fb we should use this in profiler too.) AFAICT it doesn't check nested or any other C++ subclass? (And if not, do you think it should?) Plus it includes sparse and sparse_csr, so we shouldn't need those explicit checks.

Commit:
Update on "[Autograd] Use in-place input accumulation fast path for dense Tensors."


There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

Commit:
Update on "[Autograd] Use in-place input accumulation fast path for dense Tensors."


There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

robieta commented:
I also had to gate on `var` not being a subclass because of a forward mode AD test. In theory I could have made the test check grad mode, but given how varied Tensor subclasses can be it seems like a better idea to just exclude them from the fast path. PTAL

Commit:
Update on "[Autograd] Use in-place input accumulation fast path for dense Tensors."


There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

ngimel reviewed (approved):
lgtm

Commit:
Update on "[Autograd] Use in-place input accumulation fast path for dense Tensors."


There is a fast path in InputBuffer to steal memory when use count is zero, however it is only used for sparse Tensors. According to Natalia, this is just because it wasn't obvious that there would be a benefit for dense Tensors so there was no reason to live dangerously. However I've noticed large Tensors in internal models which would benefit from this optimization as well.

Differential Revision: [D40946601](https://our.internmc.facebook.com/intern/diff/D40946601/)

cc ezyang albanD zou3519 gqchen pearu nikitaved soulitzer Lezcano Varal7

[ghstack-poisoned]

robieta commented:
@pytorchbot merge -f "test failure is unrelated. (failed to install triton)"

mehtanirav commented:
@pytorchbot revert -m "Internal test failures" -c ghfirst
